addons/website_theme_install/models/ir_module_module.py

```diff
@@ -203,43 +227,85 @@ def _unload_one_theme_module(self, website):
         (views_todel + views_todel2).unlink()
 
-    @api.multi
-    def _remove_theme_on_website(self, website):
+    def _get_upstream_themes(self):
+        """
+            get installed upstream
+        """
+        return self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
+
+    def _get_downstream_themes(self, same_name=False):
+        """
+            get installed downstream
+        """
+        return self.downstream_dependencies().filtered(lambda x: x.name.startswith(self.name if same_name else 'theme_'))
+
+    def _get_stream_themes(self):
+        """
+            Returns all the themes in the stream of the current theme.
+            First find all its downstream themes, and all of the upstream themes of both
+            sorted by their level in hierarchy, up first.
+        """
+        all_mods = self + self._get_downstream_themes(same_name=True)
+        for down_mod in self._get_downstream_themes(same_name=True) + self:
+            for up_mod in down_mod._get_upstream_themes():
+                if up_mod not in all_mods:
+                    all_mods = up_mod + all_mods
+        return all_mods
+
+    def _get_stream_website_ids(self):
+        """
+            Websites for which this theme (self) is in the stream of their theme.
+        """
+        websites = self.env['website'].search([])
+        for website in websites:
+            for theme in website.theme_id._get_stream_themes():
+                if self == theme:
+                    if website not in websites:
+                        websites += website
+        return websites
+
+    def _get_downstream_website_ids(self):
+        """
+            Websites for which this theme (self) is in the downstream of their theme.
+        """
         self.ensure_one()
-        installed_deps = self + website.theme_id.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
-        for mod in installed_deps:
-            mod._unload_one_theme_module(website)
+        websites = self.env['website'].search([])
+        for website in websites:
+            for theme in website.theme_id._get_downstream_themes():
+                if self == theme:
+                    if website not in websites:
+                        websites += website
+        return websites
 
-    @api.multi
-    def _add_theme_on_website(self, website):
-        self.ensure_one()
-        mods_to_load = reversed(self + self.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_')))
-        for mod in mods_to_load:
-            _logger.info('Load theme %s for website %s from template.' % (mod.name, website.id))
-            mod._load_one_theme_module(website, with_update=False)
-            self.env['theme.utils']._post_copy(mod)
+    def _remove_theme_on_website(self, website):
```

Comment:
in the current state, would need to be defined as api.model (or even just not be a class method since self isn't even used?)